tests/spec/puppeteer-spec.js

```diff
@@ -0,0 +1,98 @@
+
+    beforeAll(async() => {
+        // Setup a headless Chromium browser.
+        // Flags allow Puppeteer to run within a container.
+        global.browser = await global.puppeteer.launch({headless: true, args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage']});
+    });
+
+    afterAll(async() => {
+        global.browser.close();
+    });
+
+    it("Correctly loads a script from an import URL.", async function() {
+
+        const page = await global.browser.newPage();
+        const projectURL = "http://localhost:5000/editor.html?#project:XQAAgAApAQAAAAAAAAA9gn0iDP5hOXUMBZ4M1sxt7nhTa/UMRecCSq6uHLM44uVEs1hTA1G/Oa3Hy9fjvqw9MOrrvyqKstR9g9oq4yc4pkk1m9E2hvucCCCVEeUdb6bwT0S5asuGStzirbKaXcmYjTAskliKk/1v60vUCxCI/fc8ZUstwqzchTG2zAzzDii/EzhUsce8bjtDMg+OOMAzY03WeyEN6x5Z3bkVA20HbuSfofyGzVIlKfTxKeZlZVU2Wt3DdOqe1ccGelN7y0dADIpV19vKoZ9AWI8K4l3FkQQ43EIIM/vyyq0+JjpgrLhtSv/8Ma+A";
+        await page.goto(projectURL);
+
+        const codeContent = await page.evaluate("window.EDITOR.getCode();");
+        let codeName = await page.evaluate("$('#script-name').html()");
+        if (!codeName) codeName = await page.evaluate("$('#script-name').val()");
+        await page.close();
+
+        expect(codeContent).toHaveLength(111);
+        expect(codeContent).toContain("Test passed!");
+        expect(codeName).toEqual("unearthly script 2");
+
+    });
+
+    it("Shows an error dialog when loading a MakeCode hex file", async function() {
+
+        const page = await global.browser.newPage();
+        await page.goto("http://localhost:5000/editor.html");
+
+        let hasShownError = false;
+        page.on('dialog', async dialog => {
+            if (dialog.message().includes("couldn't recognise this file")) hasShownError = true;
+            await dialog.accept();
+        });
+        await page.click('#command-load');
+        await page.click('.load-drag-target.load-toggle');
+        let fileInput = await page.$('[name="load-form-file-upload"]');
+        await fileInput.uploadFile('./src/makecode.hex');
+        await page.click('[value="Load"]');
+        await page.waitFor(1000);
+        await page.close();
+
+        expect(hasShownError).toEqual(true);
+
+    });
+
+    it("Correctly loads a v1.0.1 hex file", async function() {
+
+        const page = await global.browser.newPage();
+        await page.goto("http://localhost:5000/editor.html");
+
+        await page.click('#command-load');
+        await page.click('.load-drag-target.load-toggle');
+        let fileInput = await page.$('[name="load-form-file-upload"]');
+        await fileInput.uploadFile('./src/1.0.1.hex');
+        await page.click('[value="Load"]');
+        await page.waitFor(1000);
+        const codeContent = await page.evaluate("window.EDITOR.getCode();");
+        let codeName = await page.evaluate("$('#script-name').html()");
+        if (!codeName) codeName = await page.evaluate("$('#script-name').val()");
```

Comment:
Do you know why we need to do these two lines in this order?
I take it if we remove the `$('#script-name').html()` line then the `"$('#script-name').val()`, doesn't work, is that correct?
That would indicate to me that perhaps the page for some reason needs a bit of time to process the script name. Does a delay before `let codeName = await page.evaluate("$('#script-name').val()");` "fix" the issue? (not that we should add more delays).

Reply:
This is to handle different versions of the editor. 1.1.3 uses an editable div whereas 1.1.4 uses a text input. The order doesn't matter really.

Reply:
In that case then leave what works on the master branch right now, if we change something in a different branch then we should update the tests.